docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,388 @@
+:meth:`torch.tensor()` constructor accepts :meth:`np.ndarray` objects as a data
+source to construct and return a ``torch.Tensor``.
+
+.. note::
+
+   Recall that, if ``x`` is a tensor, ``torch.tensor(x)`` is equivalent to
+    ``x.clone().detach()``.
+
+.. doctest::
+
+   >>> a_np = np.array([1, 2, 3], dtype = np.float64)
+   >>> b_torch = torch.tensor(a_np)
+   >>> b_torch
+   tensor([1., 2., 3.], dtype=torch.float64)
+   >>> b_torch[2] = 23
+   >>> np.array_equal(b_torch.numpy(), a_np)
+   False
+   >>> c_np = b_torch.numpy().copy() # new memory
+
+Calling NumPy Functions on PyTorch
+----------------------------------
+
+Operators aside, NumPy functions can be called on PyTorch tensors as well. This
+is because NumPy ``ufuncs`` or universal functions (described fully `in the
+NumPy documentation`_), take "array-like" inputs, and return tensor objects due
+to the dunder method ``__array_wrap__``.
+
+.. doctest::
+
+   >>> a_np = np.array([1, 2, 3], dtype=np.float64) / 5
+   >>> np.arctan2(a_np, 1) # No equivalent torch function
+   array([0.19866933, 0.38941834, 0.56464247])
+   >>> b_torch = torch.tensor(a_np)
+   >>> np.arctan2(b_torch, 1)
+   tensor([0.1974, 0.3805, 0.5404], dtype=torch.float64)
+
+Essentially, the code execution path is similar to the operator resolution, that is:
+
+- The PyTorch tensor is converted to a NumPy array
+- The NumPy function is executed
+- A PyTorch tensor is returned
+
+Conversely, no PyTorch functions will work on any NumPy array without explictly
+generating either a tensor `or a subclass`_.
+
+Indexing
+^^^^^^^^
+
+Indexing operations typically work as expected.
+
+.. doctest::
+
+   >>> b_torch = torch.tensor([1, 2, 3, 4, 5])
+   >>> b_torch
+   tensor([1, 2, 3, 4, 5])
+   >>> b_torch[2]
+   tensor(3)
+   >>> b_torch[-1]
+   tensor(5)
+   >>> b_torch[2:-1]
+   tensor([3, 4])
+   >>> torch.take(b_torch, torch.tensor([3, 2]))
+   tensor([4, 3])
+
+NumPy arrays can also be used for indexing.
+
+.. doctest::
+
+   >>> b_torch = torch.tensor([1, 2, 3, 4, 5])
+   >>> a_np = np.ones(3)
+   >>> b_torch[a_np]
+   tensor([2, 2, 2])
+
+Further inter-operability can be found in this `NumPy-PyTorch cheatsheet`_.
+
+.. warning::
+
+   It **is not** recommended to mix objects for indexing either.
+
+Negative strides
+~~~~~~~~~~~~~~~~
+
+NumPy arrays may have negative strides, which is not true for PyTorch tensors.
+
+.. doctest::
+
+   >>> a_np = np.array([1, 2, 3])
+   >>> b_torch = torch.from_numpy(a_np[::-1]) # doctest: +SKIP
+   Traceback (most recent call last):
+   ...
+   ValueError: At least one stride in the given numpy array is negative, and tensors with negative strides are not currently supported. (You can probably work around this by making a copy of your array  with array.copy().)
+   >>> b_torch = torch.from_numpy(np.ascontiguousarray(a_np[::-1]))
+   >>> b_torch
+   tensor([3, 2, 1])
+
+Conclusions
+-----------
+
+For the most part, implicit conversions between NumPy and PyTorch tensors are
```

Comment:
I feel like this doc is showing a lot of work to justify this statement but as a document for casual users we should probably just be clear about what we recommend/support upfront and not worry so much about proving ourselves correct